### PR TITLE
DO NOT MERGE: [constant-propagation] Re-enable compile-time overflow checks that support new integer protocols

### PR DESCRIPTION
**Pull request by swiftix**
This change implements the compile-time overflow checks that work with new integer protocols. This is achieved by:
- teaching the constant propagation pass about new generic constructors of integer types.
- It requires some very simple analysis of values stored in memory during the initialization of integers. The values go through memory, because constructors are generic now and take all arguments by reference and return their results indirectly.

Fixes rdar://problem/29937936 and rdar://problem/32158790

### CODE REVIEW

**Comment by swiftix** [comment]
@swift-ci please smoke test

**Comment by swiftix** [comment]
@moiseev Max, can you check my updates related to the tests?

**Comment by swiftix** [comment]
@eeckstein Do you mind taking a look at the copy propagation changes to support the compile-time overflow checking?

**Comment by swiftix** [comment]
@swift-ci Please test source compatibility

File: lib/SILOptimizer/Mandatory/ConstantPropagation.cpp [hunk 1]
```diff
@@ -985,10 +1032,297 @@ constantFoldStringConcatenation(ApplyInst *AI,
...
+///   %int_lit = integer_literal $Builtin.Int64, 400
+///   %int_val = struct $Int (%int_lit : $Builtin.Int64)
+///   %result = alloc_stack $UInt8
+///   %tmp = alloc_stack $Int
+///   %ty = metatype $@thick UInt8.Type
+///   store %int_val to %tmp : $*Int
+///   apply %init<UInt8, Int>(%result, %tmp, %ty) : ...
+///   dealloc_stack %tmp : $*Int
+bool ConstantPropagation::checkIntegerInitializer(
+    ApplyInst *AI, llvm::SetVector<SILInstruction *> &WorkList) {
```
Comment (eeckstein) [inline lib/SILOptimizer/Mandatory/ConstantPropagation.cpp]: This is a very big function. Is it possible to split it up?

File: lib/SILOptimizer/Mandatory/ConstantPropagation.cpp [hunk 2]
```diff
@@ -985,10 +1032,297 @@ constantFoldStringConcatenation(ApplyInst *AI,
...
+
+  // Check that this memory location is assigned just once (as an @out parameter
+  // of this initializer call).
+  bool OptimizeLoadsOfConst = isa<AllocStackInst>(ResultAddr);
+  if (OptimizeLoadsOfConst) {
+    for (auto Use : ResultAddr->getUses()) {
+      auto User = Use->getUser();
+      if (User == AI)
+        continue;
+      if (isa<StoreInst>(User) || isa<ApplyInst>(User)) {
```
Comment (eeckstein) [inline lib/SILOptimizer/Mandatory/ConstantPropagation.cpp]: I would rather use SILInstruction::mayWriteToMemory as a check here

File: lib/SILOptimizer/Mandatory/ConstantPropagation.cpp [hunk 3]
```diff
@@ -985,10 +1032,297 @@ constantFoldStringConcatenation(ApplyInst *AI,
...
+      AI->getLoc(), SILType::getPrimitiveObjectType(DstTy), {NewIL});
+
+  // Store into the ResultAddr.
+  B.setInsertionPoint(AI);
+  B.createStore(AI->getLoc(), NewStructI, ResultAddr,
+                StoreOwnershipQualifier::Unqualified);
+
+  auto RemoveCallback = [&](SILInstruction *DeadI) { WorkList.remove(DeadI); };
+
+  if (OptimizeLoadsOfConst) {
```
Comment (eeckstein) [inline lib/SILOptimizer/Mandatory/ConstantPropagation.cpp]: Is this really needed? Can this be done by subsequent mandatory passes after the constructor call is folded?

**Review by eeckstein — COMMENTED** [review]
I mostly have coding style comments. I didn't find any obvious functional problem.
Nevertheless I'm not happy with this approach. It adds much complexity and I'm wondering why all this cannot be achieved with our regular mandatory passes.

Also, if you really want to land this change you need many more tests. Basically you should have sil tests which test all of the conditions in your code - positive and negative tests.

File: include/swift/AST/ASTContext.h [hunk 4]
```diff
@@ -469,6 +470,12 @@ class ASTContext {
   /// Array.reserveCapacityForAppend(newElementsCount: Int)
   FuncDecl *getArrayReserveCapacityDecl() const;
 
+  /// Retrieve the declaration of SignedInteger.init
+  ConstructorDecl *getSignedIntegerInitDecl() const;
+
+  /// Retrieve the declaration of SignedInteger.init
```
Comment (swiftix) [inline include/swift/AST/ASTContext.h]: Oops. Will fix.

File: lib/AST/ASTContext.cpp [hunk 5]
```diff
@@ -1040,6 +1046,97 @@ FuncDecl *ASTContext::getArrayReserveCapacityDecl() const {
...
+
+      Impl.SignedIntegerInitDecl = FnDecl;
+      return FnDecl;
+    }
+  }
+  llvm_unreachable("Cannot find SignedInteger.init");
+  return nullptr;
+}
+
+ConstructorDecl *ASTContext::getUnsignedIntegerInitDecl() const {
```
Comment (swiftix) [inline lib/AST/ASTContext.cpp]: May be. I could try.

File: lib/SILOptimizer/Mandatory/ConstantPropagation.cpp [hunk 6]
```diff
@@ -985,10 +1032,297 @@ constantFoldStringConcatenation(ApplyInst *AI,
...
+///   dealloc_stack %tmp : $*Int
+bool ConstantPropagation::checkIntegerInitializer(
+    ApplyInst *AI, llvm::SetVector<SILInstruction *> &WorkList) {
+  auto &Ctx = AI->getModule().getASTContext();
+
+  assert(AI->getSubstitutions().size() == 2);
+
+
+  // First check source and destination types.
+  auto GetStdlibIntTypeInfo = [](CanType Ty, unsigned &TySize,
```
Comment (swiftix) [inline lib/SILOptimizer/Mandatory/ConstantPropagation.cpp]: Not really. I could easily move it out of the function into a separate static function.

File: lib/SILOptimizer/Mandatory/ConstantPropagation.cpp (hunk 3 above)
Comment (swiftix) [inline lib/SILOptimizer/Mandatory/ConstantPropagation.cpp]: We don't run any cleanup passes (e.g. simplifyCfg, DCE or SILCombine) at -Onone. Which other pass could perform a cleanup?

**Comment by moiseev** [comment]
@swiftix should we close this one?

**Comment by swiftix** [comment]
Yeah, I guess we can close it for now...

**Pull request closed without merging**